From 8ccb4cf61ea3439e234d747c87253d1ca7c7193d Mon Sep 17 00:00:00 2001
From: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
Date: Mon, 28 Nov 2022 11:53:02 +0530
Subject: [PATCH] ath12k: sub channel enable for dfs channel puncturing feature

The sub chan marking should be enabled in firmware to support
puncturing interference detected sub channel.

Signed-off-by: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 8 ++++++++
 drivers/net/wireless/ath/ath12k/wmi.h | 1 +
 2 files changed, 9 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -8373,6 +8373,16 @@ static int ath12k_mac_radio_start(struct
 		goto err;
 	}
 
+	/* Enable(1)/Disable(0) sub channel marking */
+	if (pdev->cap.supported_bands & WMI_HOST_WLAN_5G_CAP) {
+		ret = ath12k_wmi_pdev_set_param(ar, WMI_PDEV_PARAM_SUB_CHANNEL_MARKING,
+						1, pdev->pdev_id);
+		if (ret) {
+			ath12k_err(ab, "failed to enable SUB CHANNEL MARKING: %d\n", ret);
+			goto err;
+		}
+	}
+
 	__ath12k_set_antenna(ar, ar->cfg_tx_chainmask, ar->cfg_rx_chainmask);
 
 	/* TODO: Do we need to enable ANI? */
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -1004,6 +1004,7 @@ enum wmi_tlv_pdev_param {
 	WMI_PDEV_PARAM_RADIO_CHAN_STATS_ENABLE,
 	WMI_PDEV_PARAM_RADIO_DIAGNOSIS_ENABLE,
 	WMI_PDEV_PARAM_MESH_MCAST_ENABLE,
+	WMI_PDEV_PARAM_SUB_CHANNEL_MARKING = 0xb0,
 	WMI_PDEV_PARAM_SET_CMD_OBSS_PD_THRESHOLD = 0xbc,
 	WMI_PDEV_PARAM_SET_CMD_OBSS_PD_PER_AC = 0xbe,
 	WMI_PDEV_PARAM_ENABLE_SR_PROHIBIT = 0xc6,
